File: _scripts/templates/Dart/st.BinaryCharStream.dart
import 'package:antlr4/antlr4.dart';

import 'MyErrorListener.dart';
import 'dart:io';
import 'dart:convert';

class BinaryCharStream extends CharStream {

  late CharStream _is;

  BinaryCharStream(CharStream input_stream) : super() {
    this._is = input_stream;
  }

  @override
  int get index {
    return _is.index;
  }

  @override
  int get size {
    return _is.size;
  }

  @override
  void consume() {
    _is.consume();
  }

  @override
  int? LA(int offset) {
    return _is.LA(offset);
  }

  @override
  int mark() {
    return _is.mark();
  }

  @override
  void release(int marker)
  {
    _is.release(marker);
  }

  @override
  void seek(int _index) {
    _is.seek(_index);
  }

  @override
  String getText(Interval interval) {
    final StringBuffer buf = StringBuffer();
    final int start = interval.a;
    final int stop = interval.b;
    final int index = _is.index;
    _is.seek(0);
    for (int i = start; i \<= stop; i++) {
      final int t = _is.LA(i + 1)!;
      buf.write(t);
    }
    _is.seek(index);
    return buf.toString();
  }

  @override
  String toString() {
    return _is.toString();
  }

  @override
  String get sourceName {
    return _is.sourceName;
  }
}

File: _scripts/templates/Antlr4ng/st.BinaryCharStream.ts
// BSD License
// Copyright (c) 2020 Tom Everett
// All rights reserved.

import { CharStream, Interval } from 'antlr4ng';

export class BinaryCharStream implements CharStream {
    private stream: CharStream;

    constructor(stream: CharStream) {
        this.stream = stream;
    }

    reset(): void {
    }

    consume(): void {
        this.stream.consume();
    }

    LA(i: number): number {
        return this.stream.LA(i);
    }

    mark(): number {
        return this.stream.mark();
    }

    release(marker: number): void {
        this.stream.release(marker);
    }

    seek(index: number): void {
        this.stream.seek(index);
    }

    get name(): string {
	return this.stream.getSourceName();
    }

    get index(): number {
        return this.stream.index;
    }

    get size(): number {
        return this.stream.size;
    }

    getSourceName(): string {
        return this.stream.getSourceName();
    }

    getTextFromRange(start: number, stop: number): string {
        const buf: string[] = [];
        const index = this.stream.index;
        this.stream.seek(0);
        for (let i = start; i \<= stop; i++) {
            const t = this.stream.LA(i + 1);
            buf.push(t.toString());
        }
        this.stream.seek(index);
        return buf.join(' ');
    }

    getTextFromInterval(interval: Interval): string {
        return this.getTextFromRange(interval.start, interval.stop);
    }
}
